### PR TITLE
inliner: Check for codegen fn attributes compatibility

### LINKED ISSUE #76259: LLVM ERROR: Do not know how to split this operator's operand! with -Zmir-opt-level=2
```rust
#![crate_type = "lib"]
use std::arch::x86_64::*;

pub unsafe fn f(a: __m256i) -> u32 {
    _mm256_movemask_epi8(a).trailing_zeros()
}
```

```console
$ rustc a.rs -Zmir-opt-level=2 -C target-cpu=generic
SplitVectorOperand Op #1: t7: i32 = llvm.x86.avx2.pmovmskb TargetConstant:i64<6950>, t6

LLVM ERROR: Do not know how to split this operator's operand!
```

This looks like an interaction between MIR inliner and `#[target_feature(...)]`.
Most likely inlining shouldn't take place in-between functions with incompatible target features.

### PR DESCRIPTION
**Pull request by tmiasko**
* Check for target features compatibility
* Check for no_sanitize attribute compatibility

Fixes #76259.

### CODE REVIEW

File: compiler/rustc_mir/src/transform/inline.rs [hunk 1]
```diff
@@ -236,9 +238,15 @@ impl Inliner<'tcx> {
             return false;
         }
 
-        // Avoid inlining functions marked as no_sanitize if sanitizer is enabled,
-        // since instrumentation might be enabled and performed on the caller.
-        if self.tcx.sess.opts.debugging_opts.sanitizer.intersects(codegen_fn_attrs.no_sanitize) {
+        let self_features = &self.codegen_fn_attrs.target_features;
+        let callee_features = &codegen_fn_attrs.target_features;
+        if !callee_features.iter().all(|feature| self_features.contains(feature)) {
```
Comment (ecstatic-morse) [inline compiler/rustc_mir/src/transform/inline.rs:243]:
```suggestion
        if callee_features.iter().any(|feature| !self_features.contains(feature)) {
```
Nit: I find these kinds of conditionals easier to parse when they're in normal form (negation on the inside).

File: compiler/rustc_mir/src/transform/inline.rs [hunk 2]
```diff
@@ -236,9 +238,15 @@ impl Inliner<'tcx> {
...
+            debug!("`callee has extra target features - not inlining");
+            return false;
+        }
+
+        if self.codegen_fn_attrs.no_sanitize != codegen_fn_attrs.no_sanitize {
```
Comment (ecstatic-morse) [inline compiler/rustc_mir/src/transform/inline.rs:248]: It seems like we could do inlining when a sanitizer is enabled in one function but disabled in another so long as that same sanitizer is disabled globally. Is that correct? If so, we could do something like the following, although handling this case optimally may not be worth the effort.

```suggestion
if self.tcx.sess.opts.debugging_opts.sanitizer.iter().all(|san| self.codegen_fn_attrs.no_sanitize.contains(san) == codegen_fn_attrs.no_sanitize.contains(san)) {
```

**Review by ecstatic-morse — COMMENTED** [review]
Thanks! Left a question and a style nit.

File: compiler/rustc_mir/src/transform/inline.rs (hunk 2 above)
Comment (tmiasko) [reply compiler/rustc_mir/src/transform/inline.rs]: I limited the comparison to enabled sanitizers.

**Comment by ecstatic-morse** [comment]
Thanks!

@bors r+

**Pull request merged**